content/code-generation-from-the-ast.md

```diff
@@ -0,0 +1,287 @@
+
+We also grab the type information for our expressions and stuff it
+into a map via  https://godoc.org/golang.org/x/tools/go/types#Info .
+
+We walk the AST in depth first order via
+http://golang.org/pkg/go/ast/#Inspect . (remember, we are dealing
+w/trees)
+
+For each node we look for a few things:
+
+  * does this node contain a go routine?
+
+  * does this node make an assignment? if so is the type an error?
+
+
+## Errors
+
+  If an error is found we do a couple of things.
+
+  First, we want to mark the position in the file where we found it as
+we'll be modifying the source later. To do that we use
+http://golang.org/pkg/go/token/#FileSet.Position .
+
+  Then we grab the actual error name to have a reference to it.
+
+  We also check to see if a blank identifier, the underscore is being
+used. There are lots of times when errors get thrown away because people
+don't want to deal with them. If it's a return from strconv.Atoi maybe you don't
+care but maybe you should.
+
+  For example say you are ingesting a form request. One of the fields
+has always been a whole dollar amount so your code looks like:
+
+```go
+  s = "10"
+  i, _ := strconv.Atoi(s)
+```
+
+  However, recently, unbeknownst to you, one of the front-end people
+with code in a completely different repository on a completely different
+team dressed it up with the USD symbol of $.
+
+  Now it looks like:
+
+```go
+  s = "$10"
+  i, _ := strconv.Atoi(s)
+```
+
+  Your tests might still pass because your assumptions were that no one
+would ever send you anything but a whole dollar amount. However, if you
+look at the error you are throwing away you would know why this code is
+not working:
+
+  <u>strconv.ParseInt: parsing "$10": invalid syntax</u>
+
+  Guess what the value of i is now? That's right - [the zero value](https://golang.org/ref/spec#The_zero_value "the zero value") of an int - 0.
+
+  By discovering the places where we purposely or inadvertently throw
+away errors we can uncover potential bugs lurking on our production
+systems and prevent a rash of angry customers wondering why the service
+was free for so long when we should have been charging them $10.
+
+## Go Routines:
+
+  For go routines we simply insert code to defer any potential panics that might
+happen. For better or for worse not everyone has a "let it crash"
+mentality and sometimes it's simply not acceptable to kill the entire
+daemon because someone missed a divide by zero.
+
+By consulting the AST we can confirm what our code is doing rather than just by
+guessing via some regexen.
+
+There's a famous [stackoverflow post](http://stackoverflow.com/questions/1732348/regex-match-open-tags-except-xhtml-self-contained-tags "stackoverflow post")
+where someone asks about using regexen to parse html. Similarly, if you are trying to understand source
+- don't guess - just look it up in the AST.
+
+For an example of how we re-write the source grab deprehend && the
```

Comment:
rewrite doesn't need a hyphen
Deprehend is capitalized elsewhere.
write out && as and